javascript/ql/src/semmle/javascript/security/dataflow/MethodNameInjection.qll

```diff
@@ -0,0 +1,148 @@
+      super.isSanitizer(node) or
+      node instanceof Sanitizer
+    }
+
+    /**
+     * Holds if a property of the given object is an unsafe function.
+     */
+    predicate isUnsafeBaseObject(DataFlow::SourceNode node) {
+      // eval an friends can be accessed from the global object.
+      node = DataFlow::globalObjectRef()
+      or
+      // 'constructor' property leads to the Function constructor.
+      node.analyze().getAValue() instanceof AbstractCallable
+      or
+      // Assume that a value that is invoked can refer to a function.
+      exists (node.getAnInvocation()) 
+    }
+
+    /**
+     * Holds if the `node` is of form `Object.create(null)` and so it has no prototype.
+     */
+    predicate isPrototypeLessObject(DataFlow::MethodCallNode node) {
+      node = DataFlow::globalVarRef("Object").getAMethodCall("create") and
+      node.getArgument(0).asExpr() instanceof NullLiteral
+    }
+
+    override predicate isAdditionalFlowStep(DataFlow::Node src, DataFlow::Node dst, DataFlow::FlowLabel srclabel, DataFlow::FlowLabel dstlabel) {
+      // Reading a property of the global object or of a function
+      exists (DataFlow::PropRead read |
+        isUnsafeBaseObject(read.getBase().getALocalSource()) and
+        src = read.getPropertyNameExpr().flow() and
+        dst = read and
+        srclabel = taint() and
+        dstlabel = unsafeFunction())
+      or
+      // Reading a chain of properties from any object with a prototype can lead to Function
+      exists (PropertyProjection proj |
+        not isPrototypeLessObject(proj.getObject().getALocalSource()) and
+        src = proj.getASelector() and
+        dst = proj and
+        srclabel = taint() and
+        dstlabel = unsafeFunction())
+    }
+  }
+
+  /**
+   * A source of remote user input, considered as a source for method name injection. 
+   */
+  class RemoteFlowSourceAsSource extends Source {
+    RemoteFlowSourceAsSource() { this instanceof RemoteFlowSource }
+  }
+
+  /**
+   * The page URL considered as a flow source for method name injection. 
+   */
+  class DocumentUrlAsSource extends Source {
+    DocumentUrlAsSource() { isDocumentURL(asExpr()) }
+  }
+
+  /**
+   * A function invocation of an unsafe function, as a sink for remote method name injection.
+   */
+  class CalleeAsSink extends Sink {
+    CalleeAsSink() {
+      this = any(DataFlow::InvokeNode node).getCalleeNode()
+    }
+
+    override DataFlow::FlowLabel getFlowLabel() {
+      result = unsafeFunction()
+    }
+  }
+
+  /** 
+   * A binary expression that sanitzes a value for method name injection. That 
+   * is, if a string is prepended or appended to the remote input, an attacker 
+   * cannot access arbitrary properties.  
+   */
+  class ConcatSanitizer extends Sanitizer, DataFlow::ValueNode {
+
+    override BinaryExpr astNode;
+
+    ConcatSanitizer() {
+      astNode.getAnOperand() instanceof ConstantString      
+    }
+  }
```

Comment:
Do we also want to add comparison to constant strings as a sanitizer?

Reply:
Isn't that always a sanitizer?

Reply:
Ah, yes, you're right.